This script contains the following points:
1. Import libraries and data (df_customers & df_ords_prods_merge)
2. Exploring df_customers dataframe
3. Data wrangling of df_customers dataframe
4. Checking data consistency of df_customers
5. Merging both dataframes and exporting merged dataframe

01. Import libraries and data

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Import data
path = r'C:\Users\polly\Documents\Instacart Basket Analysis'
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)  
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_derived_more.pkl'))    

02. Exploring the data

In [3]:
# Check output of dataframe
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [4]:
# Check output of dataframe
df_customers.tail()

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095

In [5]:
# Check size of dataframe
df_customers.shape

(206209, 10)

In [6]:
# Check datatypes and memory usage
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [7]:
# Check statistics of numeric columns
df_customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

03. Data wrangling

In [9]:
# Rename columns
df_customers.rename(columns = {'First Name' : 'first_name', 'Surnam' : 'surname', 'Gender': 'gender', 'STATE' : 'state', 'Age' : 'age', 'n_dependants' : 'number_of_dependents'}, inplace = True)

In [10]:
# Check output
df_customers.head()

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependents fam_status  income  
0                     3    married  165665  
1                     0     single   59285  
2                     2    married   99568  
3                     0     single   42049  
4                     1    married   40374

In [11]:
# Change data types for numeric columns to reduce memory usage
df_customers['user_id'] = df_customers['user_id'].astype('int32')
df_customers['age'] = df_customers['age'].astype('int8')
df_customers['number_of_dependents'] = df_customers['number_of_dependents'].astype('int8')
df_customers['income'] = df_customers['income'].astype('int32')
             

In [12]:
# Check datatypes and memory usage
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  int32 
 1   first_name            194950 non-null  object
 2   surname               206209 non-null  object
 3   gender                206209 non-null  object
 4   state                 206209 non-null  object
 5   age                   206209 non-null  int8  
 6   date_joined           206209 non-null  object
 7   number_of_dependents  206209 non-null  int8  
 8   fam_status            206209 non-null  object
 9   income                206209 non-null  int32 
dtypes: int32(2), int8(2), object(6)
memory usage: 11.4+ MB


04. Checking data consistency

In [13]:
# Check statistics of numeric columns
df_customers.describe()

user_id            age  number_of_dependents         income
count  206209.000000  206209.000000         206209.000000  206209.000000
mean   103105.000000      49.501646              1.499823   94632.852548
std     59527.555167      18.480962              1.118433   42473.786988
min         1.000000      18.000000              0.000000   25903.000000
25%     51553.000000      33.000000              0.000000   59874.000000
50%    103105.000000      49.000000              1.000000   93547.000000
75%    154657.000000      66.000000              3.000000  124244.000000
max    206209.000000      81.000000              3.000000  593901.000000

Statistics seem in order except possibly maximum salary of 593,901. This is within the realm of possibility, and could be an outlier. But it's also possible that it's a mistake. Will keep this in mind for future analysis.

In [14]:
# Check for mixed data types
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

first_name


In [15]:
# Change datatype of 'first_name' column to string
df_customers['first_name'] = df_customers['first_name'].astype('str')

In [16]:
# Confirm that there are no more mixed data types
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

In [17]:
# Check for missing values in dataframe
df_customers.isnull().sum()

user_id                 0
first_name              0
surname                 0
gender                  0
state                   0
age                     0
date_joined             0
number_of_dependents    0
fam_status              0
income                  0
dtype: int64

There are no missing values in the dataframe

In [18]:
# Check for duplicate rows
df_customers[df_customers.duplicated()]

Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, date_joined, number_of_dependents, fam_status, income]
Index: []

There are no duplicate rows in the dataframe

In [19]:
# Export df_customers to 'Prepared Data' folder
df_customers.to_csv(os.path.join(path, '02 Data','Prepared Data', 'customers_checked.csv'))

05. Merge df_customers with df_ords_prods_merge


In [20]:
# Check output of df_ords_prods_merge dataframe
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    -1.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...    price_range_loc     busiest_day  busiest_day_new  \
0         Soda  ...  Mid-range product  Regularly busy   Regularly busy   
1         Soda  ...  Mid-range product  Regularly busy     Slowest days   
2         Soda  ...  Mid-range product  Regularly busy     Slowest days   
3         Soda  ...  Mid-range product      Least busy     Slowest days   
4         Soda  ...  Mid-range product      Least busy     Slowest days   

  busiest_period_of_day max_order  loyalty_flag average_price spending_flag  \
0        Average orders        10  New customer      6.367797   Low spender   
1        Average orders        10  New customer      6.367797   Low spender   
2           Most orders        10  New customer      6.367797   Low spender   
3        Average orders        10  New customer      6.367797   Low spender   
4           Most orders        10  New customer      6.367797   Low spender   

   median_days_since_prior_order order_frequency_flag  
0                           20.0     Regular customer  
1                           20.0     Regular customer  
2                           20.0     Regular customer  
3                           20.0     Regular customer  
4                           20.0     Regular customer  

[5 rows x 24 columns]

In [21]:
# Show all columns in df_ords_prods_merge
pd.set_option('display.max_columns', None)

In [22]:
# Check output again
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    -1.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day busiest_day_new busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy        Average orders         10   
1  Regularly busy    Slowest days        Average orders         10   
2  Regularly busy    Slowest days           Most orders         10   
3      Least busy    Slowest days        Average orders         10   
4      Least busy    Slowest days           Most orders         10   

   loyalty_flag  average_price spending_flag  median_days_since_prior_order  \
0  New customer       6.367797   Low spender                           20.0   
1  New customer       6.367797   Low spender                           20.0   
2  New customer       6.367797   Low spender                           20.0   
3  New customer       6.367797   Low spender                           20.0   
4  New customer       6.367797   Low spender                           20.0   

  order_frequency_flag  
0     Regular customer  
1     Regular customer  
2     Regular customer  
3     Regular customer  
4     Regular customer

In [23]:
# Remove merge flag column in df_ords_prods_merge
df_ords_prods_merge = df_ords_prods_merge.drop(columns = ['_merge'])

In [24]:
# Check size of df_ords_prods_merge dataframe
df_ords_prods_merge.shape

(32404859, 23)

In [25]:
# Check datatypes of df_ords_prods_merge
df_ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 23 columns):
 #   Column                         Dtype  
---  ------                         -----  
 0   order_id                       int32  
 1   user_id                        int32  
 2   order_number                   int8   
 3   orders_day_of_week             int8   
 4   order_time_of_day              int8   
 5   days_since_prior_order         float16
 6   product_id                     int32  
 7   add_to_cart_order              int32  
 8   reordered                      int8   
 9   product_name                   object 
 10  aisle_id                       int8   
 11  department_id                  int8   
 12  prices                         float32
 13  price_range_loc                object 
 14  busiest_day                    object 
 15  busiest_day_new                object 
 16  busiest_period_of_day          object 
 17  max_order                      int8   
 18  

In [26]:
# Merge dataframes: df_customers and df_ords_prods_merge - inner
df_ords_prods_all = df_ords_prods_merge.merge(df_customers, on = 'user_id', indicator = True)

In [27]:
# Check output of new dataframe
df_ords_prods_all.head()

order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    -1.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day busiest_day_new busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy        Average orders         10   
1  Regularly busy    Slowest days        Average orders         10   
2  Regularly busy    Slowest days           Most orders         10   
3      Least busy    Slowest days        Average orders         10   
4      Least busy    Slowest days           Most orders         10   

   loyalty_flag  average_price spending_flag  median_days_since_prior_order  \
0  New customer       6.367797   Low spender                           20.0   
1  New customer       6.367797   Low spender                           20.0   
2  New customer       6.367797   Low spender                           20.0   
3  New customer       6.367797   Low spender                           20.0   
4  New customer       6.367797   Low spender                           20.0   

  order_frequency_flag first_name surname  gender    state  age date_joined  \
0     Regular customer      Linda  Nguyen  Female  Alabama   31   2/17/2019   
1     Regular customer      Linda  Nguyen  Female  Alabama   31   2/17/2019   
2     Regular customer      Linda  Nguyen  Female  Alabama   31   2/17/2019   
3     Regular customer      Linda  Nguyen  Female  Alabama   31   2/17/2019   
4     Regular customer      Linda  Nguyen  Female  Alabama   31   2/17/2019   

   number_of_dependents fam_status  income _merge  
0                     3    married   40423   both  
1                     3    married   40423   both  
2                     3    married   40423   both  
3                     3    married   40423   both  
4                     3    married   40423   both

In [28]:
# Check frequency of new '_merge' column
df_ords_prods_all['_merge'].value_counts()

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

In [29]:
# Check size of merged dataframe
df_ords_prods_all.shape

(32404859, 33)

In [30]:
# Export new dataframe df_ords_prods_all
df_ords_prods_all.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_all.pkl'))